torchgeo/datasets/idtrees.py

```diff
@@ -383,14 +396,44 @@ def _load_geometries(self, directory: str) -> Dict[int, Dict[str, Any]]:
         features: Dict[int, Dict[str, Any]] = {}
         for path in filepaths:
             with fiona.open(path) as src:
-                for i, feature in enumerate(src):
+                for feature in src:
                     if self.split == "train":
                         features[feature["properties"]["id"]] = feature
                     # Test set task 2 has no id
                     else:
-                        features[i] = feature
+                        id = len(features) + 1
```

Comment:
Found a phantom bug here making some of the boxes be empty due to being overriden and not existing in `self.geometries`

Reply:
I'm gonna be honest, I don't really understand what `_load_geometries` does and specifically what `features` is supposed to contain. It seems like a nested dict, but what are the keys and what are the values? Also, I see that we're using `Any` and `cast` in several places, which is usually bad. Can you elaborate on this and maybe add more comments documenting it? I want to make sure you're not the only one who understands how this Dataset works.

Reply:
It's a dictionary of geojson like features (basically a dict with properties and a geometry). For the train set these geometries have unique ids inside of the shapely file. However, for the test set they have no ids, so I have to make one up. In this case I just make one that increments every time a new feature is added.

Reply:
Added more comments for clarity

Reply:
- `self.images` is a globbed list of image paths
- `self.geometries` is a dict where each entry maps a unique id to a geometry and it's properties. (test set - task 1 does not have any geometries so it's `None` for this case.
- `self.labels` is the `itc_rsFile.csv` file read into a pandas dataframe (this is only valid for the train set, so for all other cases it is `None`)

That's correct, since there are multiple shapefiles, the test set geometries were being overriden.

My initial plan was to use the `properties["indvdID"]` however I found that there were duplicates so I couldn't use this as the id so the next option was to just make a dummy id.

We could use further optimize by using a counter `i = 0` but `self._load_geometries` is only called once on instantiation of the dataset so the loop only occurs once. If you prefer I add this then that is fine with me.

Reply:
Do we need a dict for geometries? Could we use a list instead?

Reply:
The dict makes it easy for me to grab the ids and then just index the dict for the geometries needed for the image. I could revamp the dataset and just have a list of [image_path, list_of_geometries_per_image, and list_of_labels_per_image] instead?

Reply:
I think I still don't understand. If `self.images`, `self.geometries`, and `self.labels` don't share the same keys/order, how do you map from one to the other? I'm trying to follow the path from `__getitem__` index to file loading but there are a lot of moving parts.

Reply:
getitem[idx] -> image = self.images[idx]  -> self.labels[image] -> ids -> self.labels AND self.geometries

Reply:
I agree this likely needs a refactor to simplify things however, this PR is for bug fixes. The dataset currently works fine as is. Would you like to refactor for 0.4.0?

Reply:
Okay, but if `self.geometries[image]`, then why does it feel like the id in `features[id]` is completely arbitrary and not the name of the image file?

Reply:
The id is only arbitrary for the test set not the train set. You need the id to index the csv file to find the label and image filename.

Reply:
Wait, so `self._geometries()` is only called on the test set? Because it looks like the id is arbitrary any time this function is called...

Reply:
Nvm, I just noticed the if statement.

Reply:
I can follow up with a refactor in another PR. I think what I will do is on instantiation do all the ugly parsing and create a list of tuples of (image_path, boxes, box_labels) and that should be less convoluted when getitem is actually called.